### PR TITLE
Support mutually recursive traits

### LINKED ISSUE #94: Add support for mutually recursive trait definitions
Mutually recursive traits happen in practice, for example [here](https://github.com/arkworks-rs/algebra/blob/bb663bc27e776af81e66c94d9016bde2dbba9968/ff/src/fields/prime.rs#L27) (mutual recursion between `PrimeField` and `Field`):
```rust
pub trait Trait1 {
    type T: Trait2;
}

pub trait Trait2: Trait1 {}
```

It would be good to handle those, or at least print a nice error message. For now, Charon panics with:
```
thread 'rustc' panicked at 'Invalid trait decl group:
tests4::Trait1
tests4::Trait2', src/reorder_decls.rs:116:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: could not compile `tests4` (bin "tests4")
```

### PR DESCRIPTION
**Pull request by R1kM**
This PR extends support of Charon for mutually recursive trait declarations, and changes the charon-ml GAst and serialization accordingly.

In addition to the Charon support, it also adds one small example to traits.rs, and moves the unsupported recursive traits issue to supported tests.

Fixes #94

### CODE REVIEW

File: charon/tests/ui/issue-94-recursive-trait-defns.rs [hunk 1]
```diff
@@ -1,4 +1,3 @@
-//@ known-panic
 pub trait Trait1 {
     type T: Trait2;
 }
```
Comment (Nadrieril) [inline charon/tests/ui/issue-94-recursive-trait-defns.rs:3]: Could you add more examples of ways for traits to be mutually recursive? E.g. what you added to the other file, mutual reference via method signatures, via method calls, implicit reference due to parent traits, etc, as well as examples that could look recursive but aren't.

File: charon/src/reorder_decls.rs [hunk 2]
```diff
@@ -101,27 +101,22 @@ impl DeclarationGroup {
...
         gr: impl Iterator<Item = TraitDeclId::Id>,
     ) -> Self {
         let gr: Vec<_> = gr.collect();
-        let ctx = ctx.into_fmt();
         // Trait declarations often refer to `Self`, like below,
         // which means they are often considered as recursive by our
         // analysis. TODO: do something more precise. What is important
         // is that we never use the "whole" self clause as argument,
         // but rather projections over the self clause (like `<Self as Foo>::u`,
         // in the declaration for `Foo`).
```
Comment (Nadrieril) [inline charon/src/reorder_decls.rs:114]: If you understand what this TODO refers to, I'd like us to have a test for a case like that

Comment (Nadrieril) [reply charon/src/reorder_decls.rs:114]: This TODO is exactly the error we got here: https://github.com/AeneasVerif/charon/issues/159

File: charon/tests/ui/issue-94-recursive-trait-defns.rs (hunk 1 above)
Comment (R1kM) [reply charon/tests/ui/issue-94-recursive-trait-defns.rs]: I moved the tests to charon/tests/ui (as they are unsupported by Aeneas) and added a few extra tests.
As `dyn` is currently unsupported, I do not think I can add tests using method signatures, but please feel free to expand if you have other ideas of patterns to test.

**Comment by Nadrieril** [comment]
Looks good. Could you squash your commits together? Then we can merge

**Pull request merged**